pybamm/simulation.py

```diff
@@ -340,19 +342,51 @@ def solve(
             solver = self.solver
 
         if self.operating_mode == "without experiment":
-            # Solve the normal way, with a single solve
-            if t_eval is None:
-                try:
-                    # Try to compute discharge time
-                    tau = self._parameter_values.evaluate(self.model.param.timescale)
-                    C_rate = self._parameter_values["C-rate"]
-                    t_end = 3600 / tau / C_rate
-                    t_eval = np.linspace(0, t_end, 100)
-                except AttributeError:
-                    t_eval = np.linspace(0, 1, 100)
+            # For drive cycles (current provided as data) we perform additional tests
+            # on t_eval (if provided) to ensure the returned solution captures the
+            # input. If the current is provided as data then the "Current function [A]"
+            # is the tuple (filename, data).
+            if isinstance(self._parameter_values["Current function [A]"], tuple):
+                filename = self._parameter_values["Current function [A]"][0]
+                tau = self._parameter_values.evaluate(self.model.param.timescale)
+                time_data = (
+                    self._parameter_values["Current function [A]"][1][:, 0] / tau
+                )
+                # If no t_eval is provided, we use the times provided in the data.
+                if t_eval is None:
+                    pybamm.logger.info(
+                        "Setting t_eval as specified by the data '{}'".format(filename)
+                    )
+                    t_eval = time_data
+                else:
+                    # If t_eval is provided we check that it at least contains all
+                    # of the data points. We only raise a warning here, as users
+                    # may genuinely only want the solution returned at some
+                    # specified points.
+                    if set(time_data).issubset(set(t_eval)) is False:
```

Comment:
I'm wondering if this will have problems with round-off errors? And should be also allow `t_eval` with a sampling time smaller than the smallest gap in the data, but that doesn't hit exactly the time points?

Reply:
yeah good point re round-off errors. yeah that would probably be fine. not sure of the best way of testing this for all cases and give an informative message (e.g. if t_eval or the data non-uniform, would you then just check the biggest gap in t_eval and smallest in the data and just warn that the user should check to ensure they have enough resolution in t_eval?). perhaps doing that is best, as doing something naive like just checking the number of values might pass if e.g. t_eval had loads of points clustered near the start and not many later

Reply:
ah yeah I hadn't thought of the non-uniform t_eval case. I think we can allow both options: checking that either the biggest gap in t_eval is smaller than the smallest in the data, or that each point in the data has a corresponding point in t_eval within some small error tolerance, should cover most cases

Reply:
I've updated to just check max dt_eval vs min dt_data as this also catches the case where t_eval = t_data (since dt is equal here). let me know your thoughts and i'll change or merge as is

Reply:
I don't think this would  always catch the case t_eval = t_data, for example if `t_eval=t_data = [0,1,5,6]` then `max(t_eval) = 4 > 1 = min(t_data)`

Reply:
oh yeah good point haha